File: leetcode/1125_Smallest_Sufficient_Team.py
from typing import List, Dict
from unittest import TestCase
from functools import reduce

class Solution:
    def smallestSufficientTeam(self, req_skills: List[str], people: List[List[str]]) -> List[int]:
        people_n, skills_n = len(people), len(req_skills)
        skills_dic = {skill: i for i, skill in  enumerate(req_skills)}
        people_skills_bw = [reduce( lambda x, y: x | y, ( 1 << skills_dic[skill] for skill in skills), 0) for skills in people]
        for i in range(people_n):
            for j in range(i + 1, people_n):
                if people_skills_bw[i] | people_skills_bw[j] == people_skills_bw[i]:
                    people_skills_bw[j] = 0
        dp: Dict[int, int] = {0: 0}
        for person in range(people_n):
            if people_skills_bw[person] > 0:
                for skills_bw, people_bw in  list(dp.items()):
                    next_skills_bw = skills_bw | people_skills_bw[person]
                    next_people_bw = people_bw | (1 << person)
                    if next_skills_bw not in dp or bin(dp[next_skills_bw]).count('1') > bin(next_people_bw).count('1'):
                        dp[next_skills_bw] = next_people_bw
        people_bw = dp[(1 << skills_n) - 1]        
        return [i for i in range(people_n) if 1 << i & people_bw  != 0]


class SmallestSufficientTeamTest(TestCase):
    
    def test_1(self):
        self.assertEqual([0,2], sorted(Solution().smallestSufficientTeam(["java","nodejs","reactjs"], [["java"],["nodejs"],["nodejs","reactjs"]])))

    def test_2(self):
        self.assertEqual(
            [1,2], 
            sorted(Solution().smallestSufficientTeam(
                ["algorithms","math","java","reactjs","csharp","aws"], 
                [["algorithms","math","java"],["algorithms","math","reactjs"],["java","csharp","aws"],["reactjs","csharp"],["csharp","math"],["aws","java"]])
            ))

File: leetcode/1125_Smallest_Sufficient_Team.rs
use std::collections::HashMap;
struct Solution {}

impl Solution {
    pub fn smallest_sufficient_team(req_skills: Vec<String>, people: Vec<Vec<String>>) -> Vec<i32> {
        let (skills_n, people_n) = (req_skills.len(), people.len());
        let skills_dic: HashMap<String, usize> = req_skills.into_iter().enumerate().map(|(i, skill)| (skill, i)).collect();
        let mut people_skills_bw: Vec<u16> = people.into_iter()
            .map(|skills| skills.into_iter().fold(0, |acum, skill|  1 << skills_dic[&skill] | acum))
            .collect();
        for i in 0..people_n {
            for j in i + 1..people_n {
                if people_skills_bw[i] | people_skills_bw[j] == people_skills_bw[i] {
                    people_skills_bw[j] = 0;
                }
            }
        }
        let mut dp: HashMap<u16, u64> = HashMap::from([(0, 0)]);
        for person in 0..people_n {
            if people_skills_bw[person] > 0 {
                let dp_slice: Vec<(u16, u64)> = dp.iter().map(|(skills, people)| (*skills, *people)).collect();
                for (skills, people) in dp_slice {
                    let next_skills = skills | people_skills_bw[person];
                    let next_people = (1 << person) | people;
                    if ! dp.contains_key(&next_skills) || dp[&next_skills].count_ones() > next_people.count_ones() {
                        dp.insert(next_skills, next_people);
                    }
                }
            }
        }
        let people = dp[&(((1u32 << skills_n) - 1) as u16)];
        (0..people_n).filter(|person| 1 << *person & people != 0).map(|person| person as i32).collect()
    }
}

#[test]
fn test_1() {
    assert_eq!(
        vec![0,2], 
        Solution::smallest_sufficient_team(
            vec!["java".to_string(),"nodejs".to_string(),"reactjs".to_string()], 
            vec![vec!["java".to_string()],vec!["nodejs".to_string()],vec!["nodejs".to_string(),"reactjs".to_string()]]))
}

#[test]
fn test_2() {
    assert_eq!(
        vec![1,2], 
        Solution::smallest_sufficient_team(
            vec!["algorithms".to_string(),"math".to_string(),"java".to_string(),"reactjs".to_string(),"csharp".to_string(),"aws".to_string()], 
            vec![
                vec!["algorithms".to_string(),"math".to_string(),"java".to_string()],
                vec!["algorithms".to_string(),"math".to_string(),"reactjs".to_string()],
                vec!["java".to_string(),"csharp".to_string(),"aws".to_string()],
                vec!["reactjs".to_string(),"csharp".to_string()],
                vec!["csharp".to_string(),"math".to_string()],
                vec!["aws".to_string(),"java".to_string()]]))
}
